mediagrains/utils/synchronise.py

```diff
@@ -30,80 +30,132 @@ def __init__(self, *args, **kwargs):
 
     def _run_awaitable_in_new_thread(a: Awaitable[T]) -> T:
         def __inner() -> None:
             loop = asyncio.new_event_loop()
-            loop.run_until_complete(_capture_results_async(a, results))
+            loop.run_until_complete(results.capture(a))
             loop.close()
 
         t = threading.Thread(target=__inner)
         t.start()
         t.join()
 
-        return _restore_results(results)
+        return results.restore()
 
-    try:
-        loop = asyncio.get_event_loop()
-    except RuntimeError:
-        # In Python 3.6 trying to get an event loop when none can be automatically created is a an exception
-        return _run_awaitable_inside_new_runloop(f)
+    loop = get_non_running_loop()
+    if loop is not None:
+        return _run_awaitable_in_existing_run_loop(f, loop)
     else:
-        if loop.is_running():
-            # There is already a running loop on this thread so we need to spawn a new thread
-            return _run_awaitable_in_new_thread(f)
-        else:
-            # We have An existing event loop, but it is not running, so we can run it.
-            return _run_awaitable_in_existing_run_loop(f, loop)
+        return _run_awaitable_in_new_thread(f)
 
 
 def run_asyncgenerator_synchronously(gen: AsyncIterator[T]) -> Iterator[T]:
     async def __get_next(gen):
         return await gen.__anext__()
 
-    while True:
-        try:
-            yield run_awaitable_synchronously(__get_next(gen))
-        except StopAsyncIteration:
-            return
+    if get_non_running_loop() is not None:
+        while True:
+            try:
+                yield run_awaitable_synchronously(__get_next(gen))
+            except StopAsyncIteration:
+                return
+    else:
+        results = ResultsType[T]()
+
+        def _run_generator_in_new_thread(gen: AsyncIterator[T]) -> Tuple[threading.Thread, threading.Event, threading.Event]:
+            agen_should_yield = threading.Event()
+            agen_has_yielded = threading.Event()
+
+            def __inner() -> None:
+                async def _run_asyncgen_with_events(gen: AsyncIterator[T], agen_should_yield: threading.Event, agen_has_yielded: threading.Event) -> T:
+                    try:
+                        async for x in gen:
+                            agen_should_yield.wait()
+                            agen_should_yield.clear()
+                            results.rval = (x,)
+                            agen_has_yielded.set()
+
+                        agen_should_yield.wait()
+                        agen_should_yield.clear()
+                        raise StopAsyncIteration
+                    finally:
+                        agen_has_yielded.set()
+
+                loop = asyncio.new_event_loop()
+                loop.run_until_complete(results.capture(_run_asyncgen_with_events(gen, agen_should_yield, agen_has_yielded)))
+                loop.close()
+
+            t = threading.Thread(target=__inner)
```

Comment:
What happens to the thread if the generator does not reach the end, i.e. the thread doesn't complete? Is it something the garbage collector will take care of?

Reply:
Hmmm, I think maybe I need to set the `daemon` flag on the thread to make sure it gets garbage collected.